ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto

```diff
@@ -57,11 +58,37 @@ message Commands {
   // Required
   repeated Command commands = 8;
 
-  // The length of the time window during which all commands with the same ``act_as`` parties and
-  // the same command ID will be deduplicated.
-  // Duplicate commands submitted before the end of this window return an ALREADY_EXISTS error.
+  // Specifies the deduplication period for the change ID.
+  // Deduplication provides the following deduplication guarantee:
+  // If the completion for this submission is a definite answer, say at the completion offset `off`,
+  // then there is no accepting completion for the same change ID between the start of the deduplication period and `off`.
+  //
+  // The participant or Daml ledger may reject the submission with OUT_OF_RANGE
+  // if the deduplication period starts in the future (e.g., by specifying an offset that is higher than the completion end).
+  // It may reject the submission with NOT_FOUND
+  // if the deduplication period starts too early (e.g., by specifying an offset before the completion begin).
+  //
+  // If omitted, the participant chooses the start of the deduplication period at its discretion.
   // Optional
-  google.protobuf.Duration deduplication_time = 9;
+  oneof deduplication {
```

Comment:
Consider introducing a `DeduplicationMethod` message that provides space for parameters shared over the different ways of specifying the parameters. For example, whether the validity checks for the `deduplication_offset` should be skipped or not.

Reply:
I don't yet see the use case where the application wants to skip the validity checks. If those checks are skipped, the deduplication guarantee doesn't apply. Then why does it set a deduplication offset in the first place (this is anyway optional).